### ISSUE TITLE
spack load: the constraint 'spec' matches multiple packages

### ISSUE THREAD

**Issue by JusticeForMikeBrown** [spack/spack#9077]
Hi

I ran

```
user@node176[~]$ spack load -r readline@7.0%gcc@6.3.0 arch=linux-centos7-x86_64
==> Error: the constraint '['readline@7.0%gcc@6.3.0', 'arch=linux-centos7-x86_64']' matches multiple packages:
        readline@7.0%gcc@6.3.0 arch=linux-centos7-x86_64 
        readline@7.0%gcc@6.3.0 arch=linux-centos7-x86_64 
        readline@7.0%gcc@6.3.0 arch=linux-centos7-x86_64 

==> Error: In this context exactly **one** match is needed: please specify your constraints better.
```

Which is a much-needed feature.  **Thank you!**

Though how to I discern among these versions?

As you can see the only differences are hash so how do I make `spack load` choose the right one?  It would be ideal if that could be done without hash since most users are familar with loading via hash.

```

user@node176[~]$ spack find -l -L -v readline@7.0%gcc@6.3.0
==> 3 installed packages.
-- linux-centos7-x86_64 / gcc@6.3.0 -----------------------------
4pfytjk3oyixbbluigyxp2ij3kbdllje readline@7.0  rnltvzmb7ap7mf6bgarjhdcfu4kmxyph readline@7.0  h2aywb5uil33uyhq44ytmrs3wrxf5kxg readline@7.0
```

**Comment by JusticeForMikeBrown** [comment]
Also this does not seem to respect default variants.

For example with `relion` the default would be `purpose=cluster.`  Though when I do:

`spack load -r relion@2.0.3+cuda`

I get:

```
==> Error: the constraint '['relion@2.0.3+cuda']' matches multiple packages:

relion@2.0.3%gcc@4.8.5 build_type=RelWithDebInfo +cuda cuda_arch=60 +double~double-gpu+gui purpose=cluster arch=linux-centos7-x86_64
relion@2.0.3%gcc@4.8.5 build_type=RelWithDebInfo +cuda cuda_arch=60 +double~double-gpu+gui purpose=desktop arch=linux-centos7-x86_64
```

**Comment by healther** [comment]
I'm not sure how I feel about falling back on defaults here. I definitely get your point about it being inconvenient, but I feat it will create a number of subtle bugs (like having different `packages.yaml` in the $HOME/.spack leading to different packages for the same command, while everyone assumes "that cannot be the error")

You can distinguish between multiple packages by `spack find -v` and then doing `spack load /<hash>`

**Comment by JusticeForMikeBrown** [comment]
Then, at a minimum, this error message should print all of the hashes and say how to load them by hash.  Our Spack instance has many cases where this message will appear without any resolution.

Default variants should be the default.  Otherwise why are they called defaults?

**Comment by citibeth** [comment]
You are conflating two issues here: (1) identifying one or more things you want to load, and (2) ensuring dependencies are also loaded.  Plus, the example you gave has nothing to do with Spack Environments.

Spack specs are great.  But the fundamental problem with them is they are REALLY LONG, i.e. they contain all the information from the entire DAG.  If one were willing to write an entire spec after `spack load`, then there would be no problem.

Spack also has hashes, as a shorthand for the full DAG.  But hashes look like gibberish, and mean nothing to the user.  You can `spack load` a hash, but that still begs the question of, *how do you figure out which hash to load*?

What you really want to do is... specify a bunch of packages to build, and then have Spack load exactly those packages that you just built.  That is what Spack Environments does.

**Comment by JusticeForMikeBrown** [comment]
I fail to see how example has nothing to do with Spack environments.  If there are three identical specs that could satisfy a constraint, Spack should help the user identify the spec they want to load if conflicts exist.   Maybe by at least showing the hashes in the `spack load` error output.  Or by showing which spec happens to be newer than the others based upon their creation time.  Otherwise the user has to type `spack find` and look for the hash when instead `spack load` should do that automatically if conflicting specs are found.

**Comment by citibeth** [comment]
> I fail to see how example has nothing to do with Spack environments. 

The comment below has nothing to do with Spack Environments:

> Yes @citibeth but users want a "one stop shop" command to load specs into their environment. Not to be told to use spack module loads -r r@3.5.0%gcc@6.3.0+X~external-lapack for example which generates 100 lines of text.

> Maybe by at least showing the hashes in the spack load error output. 

There are currently two places in which Spack matches a user-supplied spec against existing built packages: `spack find` and `spack load`.  I think it's counter-intuitive that the two act differently.  Also... having `spack load` simply quit is clearly not the right approach.  Asking the user which one they REALLY want is a decent idea.

In many cases, people in the past have used `spack load` to load a previously-built environment.  Having `spack load` become interactive doesn't really solve any problems here.

**Comment by citibeth** [comment]
> While I love the idea of environments, they are NOT a replacement for `spack load`

I agree.  But we need some kind of `spack load` that works in a predictable manner.  "Predictable" means that it is possible/easy to build something and then load it; and the loading part will work, no matter what else has also been installed by Spack.

How about this... `spack load` (or a new `spack load --clean` would concretize the spec you give it (according to whatever configuration YAML files you have in effect), and then try to load that fully concretized spec.  At that point, it either succeeds or fails, based on whether that spec has been built.  This way, you can always build and reliably load something with `spack install foo^bar` followed by `spack load --clean foo^bar`.

Meanwhile, `spack load --dirty` would be more interactive.  It would work like `spack uninstall` or `spack find`... it matches against installed specs.  If your spec matches multiple things, it will give you an option to select which one you really mean.

**Comment by healther** [comment]
Ah I think I got your point now, let me rephrase and test that: You basically want a `spack load` behaviour that only depends on the state of the package repositories, configuration files and provided spec, but not the set of installed packages, correct?

I'm not sure how large the use case for this will still be once environments are there. For me personally having `spack load` ask me interactively which installed spec I really wanted would be sufficient as anything that goes into scripts or CI at the moment directly relies on hashes and will be migrated to environments once they are merged. 
So personally I don't care too strongly as long as `spack load` errors out when it would produce a broken environment. But I agree with @JusticeForMikeBrown it should be more helpful to the user, can you see if #9088 would have been more helpful to you?

**Comment by citibeth** [comment]
> Ah I think I got your point now, let me rephrase and test that: You basically want a spack load behaviour that only depends on the state of the package repositories, configuration files and provided spec, but not the set of installed packages, correct?

I hadn't thought of it in those terms, but yes.  One big benefit of Spack is it can install multiple versions of the same thing, and support multiple projects, without duplicating installations that are the same between the two.  The assumption needs to be that other people or projects might be installing stuff into the Spack repo that you can't control; and those installations should not break what you are doing.

> I'm not sure how large the use case for this will still be once environments are there.

Once environments get merged (@scheibelp ???) I want to make them integral to every Spack command, rather than having them be a separate command.  (Don't worry, this won't affect existing usage).  The idea is that every command should do the natural thing when asked to do its thing "with respect to an environment."  For example, "spack install" with respect to an environment should build the package and add it to the environment.  From this point of view, the "spack load" behavior I suggested comes very naturally as "spack load" with respect to an environment.  I think people will use it and like it; because it will provide the predictability of Spack environments without requiring people to do a lot of set-up work currently required for environments.

> anything that goes into scripts or CI at the moment directly relies on hashes and will be migrated to environments once they are merged.

To be clear... if you know the hash, you don't need Spack at all, you just use `module load`.  So Spack's purpose in loading modules really seems to be to convert specs to hashes .

> can you see if #9088 would have been more helpful to you?

I would recommend merging it, printing hashes is always a good thing that we won't regret.  But I think in the long run, we need a better solution to the core problems here.

**Comment by scheibelp** [comment]
@JusticeForMikeBrown @healther @citibeth now that #9088 is merged I am going to close this, as I think that addresses the problem as stated in https://github.com/spack/spack/issues/9077#issue-353133611. Do you think it would be worthwhile to either

* open a new issue targeting the additional discussion here
* updating the text in https://github.com/spack/spack/issues/9077#issue-353133611 and reopening

? I'd lean toward the first option.

> Once environments get merged (@scheibelp ???)

Myself and @tgamblin are on it.
